src/mlpack/core/data/string_encoding_impl.hpp

```diff
@@ -0,0 +1,175 @@
+    std::deque<std::string>::iterator jt = originalStrings.begin();
+    for (auto it = oldObject.originalStrings.begin();
+        it != oldObject.originalStrings.end(); it++)
+    {
+      mappings[*jt] = oldObject.mappings.at(*it);
+      jt++;
+    }
+  }
+  return *this;
+}
+
+template<typename TokenizerType>
+void StringEncoding::CreateMap(std::string& input,
+    TokenizerType tokenizer)
+{
+  boost::string_view strView(input);
+  boost::string_view token;
+  token = tokenizer(strView);
+  std::size_t curLabel = mappings.size() + 1;
+  while (!token.empty())
+  {
+    if (mappings.find(token) == mappings.end())
+    {
+        originalStrings.push_back(std::string(token));
+        mappings[originalStrings.back()] = curLabel++;
+    }
+    token = tokenizer(strView);
+  }
+}
+
+template<typename MatType, typename TokenizerType, typename EncodingPolicy>
+void StringEncoding::Encode(
+    const std::vector<std::string>& input,
+    MatType& output, TokenizerType tokenizer,
+    EncodingPolicy policy)
+{
+  boost::string_view strView;
+  boost::string_view token;
+  std::vector< std::vector<boost::string_view> > dataset;
+  size_t colSize = 0;
+  for (size_t i = 0; i < input.size(); i++)
+  {
+    strView = input[i];
+    token = tokenizer(strView);
+    dataset.push_back(std::vector<boost::string_view>() );
+    while (!token.empty())
+    {
+      dataset[i].push_back(token);
+      token = tokenizer(strView);
+    }
+    colSize = std::max(colSize, dataset[i].size());
+  }
+  size_t curLabel = mappings.size() + 1;
+  for (size_t i = 0; i < dataset.size(); ++i)
+  {
+    for (size_t j = 0; j < dataset[i].size(); ++j)
+    {
+      if (mappings.find(dataset[i][j]) == mappings.end())
+      {
+        originalStrings.push_back(std::string(dataset[i][j]));
+        mappings[originalStrings.back()] = curLabel++;
+      }
+    }
+  }
+  EncodingPolicy::InitMatrix(output, dataset.size(), colSize, mappings.size());
+  for (size_t i = 0; i < dataset.size(); ++i)
+    for (size_t j = 0; j < dataset[i].size(); ++j)
+        EncodingPolicy::Encode(mappings.at(dataset[i][j]), output, i, j);
+  }
+
+template<typename TokenizerType,typename EncodingPolicy>
+void StringEncoding::Encode(
+    const std::vector<std::string>& input,
+    std::vector<std::vector<size_t>>& output,
+    TokenizerType tokenizer,
+    EncodingPolicy policy,
+    typename std::enable_if<PolicyTraits<EncodingPolicy>::
+    outputWithNoPadding>::type*)
+{
+  boost::string_view strView;
+  boost::string_view token;
+  size_t curLabel = mappings.size() + 1;
+  for (size_t i = 0; i < input.size(); ++i)
+  {
+    output.push_back(std::vector<size_t>() );
+    strView = input[i];
+    token = tokenizer(strView);
+    while (!token.empty())
+    {
+      if (mappings.count(token) == 0)
+      {
+        originalStrings.push_back(std::string(token));
+        mappings[originalStrings.back()] = curLabel++;
+      }
+      output[i].push_back(mappings.at(token));
```

Comment:
Looks like the encoding algorithm doesn't depend on the policy at all. This overload performs the dictionary encoding algorithm. Should we change this line by some policy function?

Reply:
May be that would be a good idea.